shared/utils/domains.ts

```diff
@@ -1,73 +1,52 @@
 import { trim } from "lodash";
+import env from "../env";
 
 type Domain = {
-  tld: string;
-  subdomain: string;
+  subdomain?: string;
   domain: string;
+  custom?: boolean;
 };
 
+// strips protocol and whitespace from input
+// then strips the path and query string
+function normalizeUrl(url: string) {
+  return trim(url.replace(/(https?:)?\/\//, "")).split(/[/:?]/)[0];
+}
+
 // we originally used the parse-domain npm module however this includes
 // a large list of possible TLD's which increase the size of the bundle
 // unnecessarily for our usecase of trusted input.
-export function parseDomain(url?: string): Domain | null | undefined {
-  if (typeof url !== "string") {
-    return null;
-  }
-  if (url === "") {
+export function parseDomain(url: string): Domain | null | undefined {
+  if (!url) {
     return null;
   }
 
-  // strip extermeties and whitespace from input
-  const normalizedDomain = trim(url.replace(/(https?:)?\/\//, ""));
-  const parts = normalizedDomain.split(".");
+  const normalBaseUrl = normalizeUrl(env.URL);
+  const normalUrl = normalizeUrl(url);
 
-  // ensure the last part only includes something that looks like a TLD
-  function cleanTLD(tld = "") {
-    return tld.split(/[/:?]/)[0];
+  const baseUrlStart = normalUrl.indexOf(`.${normalBaseUrl}`);
+  if (baseUrlStart === -1) {
+    return { subdomain: "", domain: normalUrl, custom: true };
   }
 
-  // simplistic subdomain parse, we don't need to take into account subdomains
-  // with "." characters as these are not valid in Outline
-  if (parts.length >= 3) {
-    return {
-      subdomain: parts[0],
-      domain: parts[1],
-      tld: cleanTLD(parts.slice(2).join(".")),
-    };
-  }
+  // we consider anything in front of the base url to be the subdomain
```

Comment:
🍖 this is the core of the changes, we're making the calculation of subdomain and domain very specific to Outline

Reply:
I'm down with this approach, I think the only thing that's rubbing me the wrong way is retaining the generic naming of methods and the filename. It seems inevitable that these methods will be accidentally misused down the line… 

the main one is `subdomain`, maybe we make it `teamSubdomain` ? Or we need to make `parseDomain` a method that isn't exposed as you put in the TODO

Reply:
I think `parseTeamDomain`, or `TeamDomain.parse(url).subdomain` could do the trick

Reply:
I've had similar thoughts, it's all very overloaded making this hard to name correctly.
I think the way to go is to include all the relevant properties, and try to make them all as semantically correct as possible (kind of like the built in URL class)
```
type OutlineDomain = {
  teamSubdomain: string; // only the team subdomain or empty string, never `www` or `app` 
  host: string; // including subdomain, without protocol or path... e.g. www.getoutline.com, app.getoutline.com, myteam.getoutline.com, customdomain.com
  custom: boolean;
}
```
The insight is that we done really need the domain segment just by itself. If we ever need to examine it, it's in conjunction with subdomain as well, so we're comparing the entire host string.